keptn/keptn
# feat(shipyard-controller): Leader election

## This PR


- adds k8s leader election to shipyard-controller
- unit test
- integration-test

### Related Issues


Closes #6822 

### Notes

Enabling leader election trough k8s in openshift would require a  downgraded k8s.io from v0.22.6 to around v0.0.0-20180608153232-46b5520a878b this is not a good idea... So for now Leader election can be turned off via helm value

### Follow-up Tasks



### How to test

with current graceful shutdown leader election is delayed between 30-60 sec due to graceperiod and prestophook in distributor. This means that when one leader receives a SIGTERM it will take that time before a new one is elected. You can reproduce with a single copy of the shipyard-controller, delete it and see that while the other is spawned the leadership is to a different copy, finally a new election happens.
![image](https://user-images.githubusercontent.com/89971034/154484513-84427145-3c9e-418c-8cc5-4a165edbed1a.png)
